src/frequenz/sdk/actor/power_distributing/power_distributing.py

```diff
@@ -62,6 +63,34 @@ class _User:
     """The bidirectional channel to communicate with the user."""
 
 
+@dataclass
+class _CacheEntry:
+    """Represents an entry in the cache with expiry time."""
+
+    inv_bat_pair: InvBatPair
+    """The inverter and adjacent battery data pair."""
+
+    def __init__(
+        self, inv_bat_pair: InvBatPair, duration: timedelta = timedelta(hours=2.5)
+    ) -> None:
+        """Initialize a CacheEntry instance.
+
+        Args:
+            inv_bat_pair: the inverter and adjacent battery data pair to cache.
+            duration: the duration of the cache entry.
+        """
+        self.inv_bat_pair = inv_bat_pair
+        self._expiry_time = datetime.now(tz=timezone.utc) + duration
+
+    def is_expired(self) -> bool:
+        """Check whether the cache entry has expired.
+
+        Returns:
+            whether the cache entry has expired.
+        """
+        return datetime.now(tz=timezone.utc) >= self._expiry_time
```

Comment:
For me this is still weird, having a `@dataclass` with an `__init__` method smells. Also this should ideally probably use monotonic time.

What about this?

```suggestion
@dataclass
class _CacheEntry:
    """Represents an entry in the cache with expiry time."""

    inv_bat_pair: InvBatPair
    """The inverter and adjacent battery data pair."""
    
    expiry_time: int
    """The expiration time (taken from the monotonic clock) of the cache entry."""

    @classmethod
    def from_ttl(
        cls, inv_bat_pair: InvBatPair, ttl: timedelta = timedelta(hours=2.5)
    ) -> Self:
        """Initialize a CacheEntry instance from a TTL (Time-To-Live).

        Args:
            inv_bat_pair: the inverter and adjacent battery data pair to cache.
            ttl: the time a cache entry is kept alive.
        """
        return cls(inv_bat_pair, time.monotonic_ns() + ttl)

    @property
    def has_expired(self) -> bool:
        """Check whether the cache entry has expired.

        Returns:
            whether the cache entry has expired.
        """
        return time.monotonic_ns() >= self.expiry_time
```

Reply:
This is why I was suggesting doing the cache expiry in a separate PR as well. :-)

But I guess we can wait to get it right here already.

Reply:
That looks indeed more solid and reliable. Will update it as it is suggested. Thanks!
(I didn't know we can add `timedelta` to time monotonic)

Reply:
I don't think we can, I didn't test it :laughing: , you should probably convert the `timedelta` using `int(delta.total_seconds() * 1e9)`.

But you can also do this in a follow-up PR if it is blocking @shsms, this is just really mostly stylistic.

Reply:
I'm also using `monotonic_ns` out of paranoia after having issues with `float`s when implementing the new `Timer` in the channels repo.

Reply:
I see. I've updated in the same line